src/main/kotlin/org/rust/ide/typing/paste/RsConvertJsonToStructCopyPasteProcessor.kt

```diff
@@ -0,0 +1,272 @@
+
+        val data = values.getOrNull(0) as? PotentialJsonTransferableData ?: return
+        val text = data.text
+
+        val structs = extractStructsFromJson(text) ?: return
+        if (!shouldConvertJson(project)) return
+
+        val factory = RsPsiFactory(project)
+        val nameMap = generateStructNames(structs)
+
+        val psiDocumentManager = PsiDocumentManager.getInstance(project)
+        val insertedItems: MutableList<SmartPsiElementPointer<RsStructItem>> = mutableListOf()
+
+        runWriteAction {
+            // Delete original text
+            editor.document.deleteString(bounds.startOffset, bounds.endOffset)
+            psiDocumentManager.commitDocument(editor.document)
+
+            val file = psiDocumentManager.getPsiFile(editor.document) as? RsFile ?: return@runWriteAction null
+            val element = file.findElementAt(caretOffset)
+
+            val parent = element?.parent ?: file
+            var anchor = element
+
+            for (struct in structs) {
+                val inserted = createAndInsertStruct(factory, anchor, parent, struct, nameMap) ?: continue
+                anchor = inserted
+                insertedItems.add(inserted.createSmartPointer())
+            }
+        }
+
+        if (insertedItems.isNotEmpty()) {
+            replacePlaceholders(editor, insertedItems, nameMap)
+        }
+    }
+}
+
+private fun shouldConvertJson(project: Project): Boolean {
+    return if (isUnitTestMode) {
+        true
+    } else {
+        showYesNoDialog(
+            RsBundle.message("copy.paste.convert.json.to.struct.dialog.title"),
+            RsBundle.message("copy.paste.convert.json.to.struct.dialog.text"),
+            project,
+            icon = Messages.getQuestionIcon()
+        )
+    }
+}
+
+private fun generateStructNames(structs: List<Struct>): Map<Struct, String> {
+    return if (structs.size == 1) {
+        mapOf(structs[0] to "Struct")
+    } else {
+        structs.mapIndexed { index, struct -> struct to "Struct${index + 1}" }.toMap()
+    }
+}
+
+/**
+ * Creates a PSI struct from the given Struct datatype description and inserts it after the given anchor.
+ */
+private fun createAndInsertStruct(
+    factory: RsPsiFactory,
+    anchor: PsiElement?,
+    parent: PsiElement,
+    struct: Struct,
+    nameMap: Map<Struct, String>
+): RsStructItem? {
+    val structPsi = generateStruct(factory, struct, nameMap) ?: return null
+
+    val inserted = if (anchor == null) {
+        parent.add(structPsi)
+    } else {
+        parent.addAfter(structPsi, anchor)
+    } as RsStructItem
+
+    return inserted
+}
+
+fun generateStruct(factory: RsPsiFactory, struct: Struct, nameMap: Map<Struct, String>): RsStructItem? {
+    val structString = buildString {
+        append("struct ${nameMap[struct]} {\n")
+
+        for (field in struct.fields) {
+            val name = field.key.escapeIdentifierIfNeeded()
```

Comment:
I guess we need to convert name to `snake_case` here. And probably in future PRs add `#[serde(rename_all = "...")]` based on names style in pasted JSON